### ISSUE TITLE
Windows minion and state.highstate misbehavior 

### ISSUE THREAD

**Issue by dsumsky** [saltstack/salt#20836]
I noticed a strange behavior when calling state.highstate (and potentically any other) function on Windows minion. It seems that Windows minion does not control how many instances of state.highstate function is running on minion. Let's describe my testing environment  and procedure how to reproduce it (tested on SaltStack 2014.1.13 and also  on the latest 2014.7.1):

- I have a minion accepted by a master, the master has  just default base environment configured with this top.sls file:
```
base:
  '*':
    - run_me
```
- the state file run_me.sls contains:
```
{% if grains['kernel'] == 'Windows' %}
run_me_start:
  cmd.run:
    - name: echo start >> c:\run_me.txt

run_me_sleep:
  cmd.run:
    - name: timeout 90

run_me_end:
  cmd.run:
    - name: echo end >> c:\run_me.txt & echo[ >> c:\run_me.txt
{% else %}
run_me_start:
  cmd.run:
    - name: echo start >> /run_me.txt

run_me_sleep:
  cmd.run:
    - name: sleep 90

run_me_end:
  cmd.run:
   - name: echo end >> /run_me.txt
{% endif %}
```
- basically, the states are pretty stupid and  their only purpose is to check if they overlap when they are tried to be executed twice. There is a pause of 90s in their execution but they will be called in shorter time.
- then, I have configured salt scheduler which calls state.highstate every 30s so it will be  called multiple times before the previous one ends (30s schedule < 90s pause). The configuration is distributed from pillars and it works perfectly (maxrunning option is set to 1 by default):
```
schedule:
  highstate:
    function: state.highstate
    seconds: 30
```
- when I start a minion service (default configuration from installation) on Linux (CentOS 6.5 x86_64) I can see in /run_me.txt file this sequence of execution (without overlaps):
```
start
end
start
end
```
- while on Windows (Windows server 2012 R2), I can see overlaps which make me pretty nervous:
```
start
start
end
start
end
```
- ok, when I change the schedule configuration to something like this:
```
schedule:
  highstate:
    function: cmd.run
    seconds: 30
    args:
      - 'salt-call state.highstate'
```
- it behaves on Linux the same way like in previous example:
```
start
end
start
end
``` 
- on Windows, it is better now:
```
start
end
start
end
```
- ok, so what happens when I log in to the minion, there is 'paused' highstate running (it takes 90s to execute) and I try to execute salt-call state.highstate manually?
  - on Linux where I do most of my work it ends with error as expected as I can't call it twice:
```
local:
    Data failed to compile:
----------
    The function "state.highstate" is running as PID XXXX and was started atDATE with jid JID
```
  - on Windows, it executes without any problems!!! So when I check the run_me.txt file I can  see again overlapping highstates:
```
start
start
end
end
```
- I also tried to disable salt scheduler and to configure startup_states: highstate on the minion
- then, on Linux it fine well and again, only one instance of state.highstate is allowed
- but on Windows, I'm allowed to call state.highstate as many times as I wish :-(

Can anybody explain we what am I doing wrong? Of course, I should ensure that when state.highstate is scheduled like above that nobody will call it manually. But you still need to remember how to schedule it. And why it works on Linux as expected but not on Windows?

**Comment by eliasp** [comment]
This would explain weird things happening recently on my Windows minions - I suspected something like this, but never looked that closely at it as you did! Thanks for the thorough investigations!

**Comment by rallytime** [comment]
Thanks for this detailed report @dsumsky. I'm not entirely sure what is happening here, but these details are a great start.

ping @shanedlee and @UtahDave

**Comment by dsumsky** [comment]
Is there any progress in this matter? Are you able to reproduce the bug? Further, do you have any idea how to try to workaround it?

**Comment by dsumsky** [comment]
Hello,
I ran the test with SaltStack 2014.7.5 again and the issue is still there. Any idea or suggestion how to workaround it at least?

Thanks.

**Comment by UtahDave** [comment]
I noticed yesterday while investigating another bug that the proc dir isn't
working correctly on Windows. Once that is fixed, this error you're seeing
should go away as well

**Comment by dsumsky** [comment]
Could you please provide some more details what is not working with proc dir on Windows correctly?!? Is there any plan when it could be fixed?

**Comment by twangboy** [comment]
@UtahDave ^^^^

**Comment by davidcop** [comment]
Any updates on this issue????

**Comment by dsumsky** [comment]
The bug is still present in 2015.5.3 release. It is possible to workaround it with Windows built-in scheduler which allows to schedule execution of ```state.highstate``` function and to not start a new instance of the task if it is already running. But this is applicable only under assumption if you don't need to communicate with minions from masters and so you can afford to stop/disable ```salt-minion``` service on minions.

**Comment by OpusX** [comment]
Until the bug is fixed, I wrote a work around in powershell.  It basically looks at the processes running to determine whether or not a highstate is currently running.

1) Get processes that are running and determine if the commandline associated with python.exe contains the string "state.highstate"  This will get any highstates running locally from commandline, from scheduled task, etc.
2) Get Child processes for the process python.exe whose commandline equals "c:\salt\bin\python.exe"" c:\salt\bin\Scripts\salt-minion -c c:\salt\conf -l quiet
This will get any processes spawned from a highstate initiated from SaltMaster.  

If any other processes are found you know the highstate is currently running and can use an exit 1 in your state.

I stuck the above script in an onlyif and set to order 0.  if onlyif passed the script would issue exit 1 killing the highstate.

It is sloppy but it works for now as a work around to the issue I am experiencing.

**Comment by twangboy** [comment]
@dsumsky This has been fixed on versions of salt >= 2015.8. With multiprocessing off, Windows doesn't handle the jobs correctly. Multiprocessing has been fixed and is the default on 2015.8 and newer. If you upgraded from an older version, you may need to go into the minion config and set `multiprocessing: True`
